### PR TITLE
Proposal: Add `path` column to categories table as `ltree` type to make tree traversing easier

### PR DESCRIPTION
**Pull request by erewok**
## Description

This is a proposal inspired by tickets ([1](https://github.com/rust-lang/crates.io/issues/1093), and [2](https://github.com/rust-lang/crates.io/issues/721)) with the goal of eventually revealing more information about categories on pages where category information is available. It is also motivated by the current application code surrounding categories, which contains lots of SQL with operations like this: `split_part(c2.slug, '::', 1)`.

This PR is intended as a suggestion for what to do with the `categories` table.

Following are changes included here:

- Adds a `path` column as a Postgresql `ltree` type to the `categories` table in order to more easily query trees of categories and subcategories. This allows navigating trees of arbitrary depth and potentially more ergonomic path-finding queries.

- Adds a `parent_categories` method to `Category`, which returns parent categories _in order of traversal_ from root to this node.

### Pros

- Easily handle trees of arbitrary depth, ex: "category::sub-category::sub-catgeory-A::sub-category-A-1".
- Makes queries to find a whole tree of nodes simpler (and possibly quicker?). Ex:
  `select slug, path from categories where path <@ 'root.web_programming';`
- Makes finding all root-level nodes simpler (assuming one is cool with the syntax). Ex:
  `select slug, path from categories where path ~ 'root.*{1}';`
- Makes finding the direct path from the root node to a particular node easier (see `parent_categories` method). Ex:
  `select slug, path from categories where path @> 'root.web_programming.http_client' order by path;`
- Invisible to current application code: nothing needs to be updated to handle the `path` column because it happens via database triggers. (This could also be a _con_ depending on one's perspective.)

### Cons

- Postgresql only (harder to switch databases in the future)
- Makes certain characters in slugs disallowed: `[',",-]`. Path labels must consist of "A-Za-z0-9_" separated by "." We swap `-` for `_` and `::` for `.` in the included Postgresql procedure.
- This `path` column is nowhere in application code, so this is all hand-written SQL (but the current solution is doing that as well).
- Error messages are opaque if you happen to have a bad character in a slug:
```
cargo_registry=# update categories set slug = 'test-new::bla$bla-::parent::child' where slug = 'test-new::parent::child';
ERROR:  syntax error at position 17
CONTEXT:  PL/pgSQL function set_category_path_to_slug() line 3 at assignment
```

## Notes

- [x] Needs a test for the `parent_categories` method.

### CODE REVIEW

File: src/category.rs [hunk 1]
```diff
@@ -161,6 +162,24 @@ impl Category {
         ).bind::<Text, _>(&self.category)
             .load(conn)
     }
+
+    pub fn parent_categories(&self, conn: &PgConnection) -> QueryResult<Vec<Category>> {
```
Comment (vignesh-sankaran) [inline src/category.rs]: Could you add a doc comment here outlining what this function does??

Comment (erewok) [reply src/category.rs]: You know I thought I had done that but I forgot about the three-slashes thing. Sorry about that.

**Comment by kivikakk** [comment]
I've now adapted #1130 into this PR! Parent category traversal is now easier (and accounts for arbitrary category tree depth).

**Comment by carols10cents** [comment]
Sorry this has taken me so long to look at-- my first impression is WHERE HAS LTREE BEEN ALL MY LIFE?!!?!? Like seriously, is this new? If not, I am irrationally angry at everyone i've interacted with who has known about this feature and not told me about it. I'm now going to start telling people as soon as I meet them-- "Hi, I'm Carol, have you heard about Postgres' ltree?"

Digging into the code and trying this out now :)

**Comment by erewok** [comment]
@carols10cents `ltree` has been part of Postgresql since at least 2004, I think. I only found out about in the last couple of years and I have been surprised how rarely people recommend it or even know about it because it seems like a surprisingly useful data type. My theory is that Postgresql offers so many different options that it's difficult to actually know about all of them. For instance, I'm only aware of a handful of the myriad index types and I feel like I don't hear mention of those either.

The syntax, perhaps, takes a bit to get used to, but I find it pretty ergonomic myself: 

`@> some_path`: means a path that's greater than or equal to *some_path*.
`<@ some_path`: means a path that's less than or equal to *some_path*.
`~ some_path`: looks sort of like a regex comparison, usually?

In this PR, I added `root.` to all the constructed paths, and this makes path-finding a whole lot easier as long as no one ever adds a new category with a slug called `root`!

File: migrations/2017-10-08-193512_category_trees/up.sql [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+UPDATE categories
+    SET path =  text2ltree('root.' || trim(replace(replace(slug, '-', '_'), '::', '.')))
+    WHERE path is NULL;
+
+ALTER TABLE CATEGORIES
+    ALTER COLUMN path SET NOT NULL;
+
+-- Create some indices that allow us to use GIST operators: '@>', etc
+CREATE INDEX path_gist_categories_idx ON categories USING GIST(path);
+CREATE INDEX path_categories_idx ON categories USING btree(path);
```
Comment (carols10cents) [inline migrations/2017-10-08-193512_category_trees/up.sql]: Is there a reason to have both a gist and a btree index? The [postgres docs](https://www.postgresql.org/docs/9.6/static/ltree.html) say:

> F.21.3. Indexes
> ltree supports several types of indexes that can speed up the indicated operators:
>
> * B-tree index over ltree: <, <=, =, >=, >
> * GiST index over ltree: <, <=, =, >=, >, @>, <@, @, ~, ?

which looks like gist indexes take care of the same operators that btree indexes take care of? 

All of this is probably moot anyway since the categories table is so smalll :)

File: app/templates/category/index.hbs [hunk 3]
```diff
@@ -1,8 +1,11 @@
 {{ title category.category ' - Categories' }}
 
 <div id='crates-heading'>
-    {{svg-jar "crate"}}
-    <h1>{{ category.category }}</h1>
+    {{#link-to "categories"}}{{svg-jar "crate"}}{{/link-to}}
```
Comment (carols10cents) [inline app/templates/category/index.hbs]: Neat easter-egg :)

File: migrations/2017-10-08-193512_category_trees/up.sql (hunk 2 above)
Comment (erewok) [reply migrations/2017-10-08-193512_category_trees/up.sql]: Honestly, I've just seen everyone do it that way, so I figured it was a best practice, if not necessary.

**Comment by erewok** [comment]
thanks @carols10cents and thanks to @kivikakk for working on the frontend.

One thing I considered was rewriting the other queries that do `split("::", 1)` in order for them to be all be consistent. I didn't do that because I wasn't sure if there would be buy-in for these changes. I knew there would be an issue with diesel, and I didn't want to rock the boat too much, even though I felt like this would be a useful direction to go in.

**Comment by sgrif** [comment]
You can just delete the `path` column from `schema.rs`, or create `struct Ltree;` somewhere and `use` it at the top of the table declaration.

**Comment by kivikakk** [comment]
It looks like `print-schema` won't automatically do what we want with extensions (going by this and #1133).

I've created a super basic `diesel_ltree` crate with an `Ltree` struct that does what we want (i.e. almost nothing right now), and manually added it to `schema.rs` in b7c49f8e9f.

**Comment by kivikakk** [comment]
@erewok 

> One thing I considered was rewriting the other queries that do `split_part(some_value, "::", 1)` in order for them to be all be consistent. I didn't do that because I wasn't sure if there would be buy-in for these changes. I knew there would be an issue with diesel, and I didn't want to rock the boat too much, even though I felt like this would be a useful direction to go in.

[`diesel_ltree`](https://crates.io/crates/diesel_ltree) now supports the common operators and functions on ltrees/lqueries, so we could indeed redo those idiomatically from here on out.

**Comment by sgrif** [comment]
@kivikakk `diesel_ltree` needs to use http://docs.diesel.rs/diesel/pg/struct.PgMetadataLookup.html#method.lookup_type for its `HasSqlType` impls. Types from extensions *do not* have stable OIDs. I suspect that the only reason this isn't breaking things is because you aren't currently dealing with bind params of those types

**Comment by kivikakk** [comment]
Using latest ltree which fixes this. :+1: Thanks sgrif.

**Comment by kivikakk** [comment]
@carols10cents This is brought up to speed with `master`; would you mind taking another look? Just got an accessibility violation to clear up.

File: src/category.rs [hunk 4]
```diff
@@ -156,6 +157,26 @@ impl Category {
...
+    /// Gathers the parent categories from the top-level Category to the direct parent of this Category.
+    /// Returns categories as a Vector in order of traversal, not including this Category.
+    /// The intention is to be able to have slugs or parent categories arrayed in order, to
+    /// offer the frontend, for examples, slugs to create links to each parent category in turn.
+    pub fn parent_categories(&self, conn: &PgConnection) -> QueryResult<Vec<Category>> {
+        use diesel::sql_types::Text;
+
+        sql_query(
+            "SELECT c.id, c.category, c.slug, c.description, \
+            COALESCE(( \
```
Comment (sgrif) [inline src/category.rs]: Is this `COALESCE` actually doing anything?

File: src/category.rs [hunk 5]
```diff
@@ -156,6 +157,26 @@ impl Category {
...
+
+    /// Gathers the parent categories from the top-level Category to the direct parent of this Category.
+    /// Returns categories as a Vector in order of traversal, not including this Category.
+    /// The intention is to be able to have slugs or parent categories arrayed in order, to
+    /// offer the frontend, for examples, slugs to create links to each parent category in turn.
+    pub fn parent_categories(&self, conn: &PgConnection) -> QueryResult<Vec<Category>> {
+        use diesel::sql_types::Text;
+
+        sql_query(
+            "SELECT c.id, c.category, c.slug, c.description, \
```
Comment (sgrif) [inline src/category.rs]: Can we move this whole thing into a `.sql` file and use `include_str!`, add some more newlines and indentation? The whole thing is very hard to parse. It's not obvious where the call to `COALESCE` ends for example

File: src/category.rs [hunk 6]
```diff
@@ -398,4 +424,30 @@ mod tests {
         let expected = vec![("Cat 3".to_string(), 6), ("Cat 1".to_string(), 3)];
         assert_eq!(expected, categories);
     }
+
+    #[test]
+    fn category_parent_categories_includes_path_to_node_with_count() {
+        let conn = pg_connection();
+        conn.batch_execute(
```
Comment (sgrif) [inline src/category.rs]: Can we use Diesel's public API for this (I need to update the other tests). You can write this as

```
insert_into(categories)
    .values(vec![
        (category.eq("Cat 1"), slug.eq("cat1"), crates_cnt.eq(1)),
        ...
    ])
```

File: src/lib.rs [hunk 7]
```diff
@@ -5,7 +5,7 @@
 //! [krate](krate/index.html), [user](user/index.html) and [version](version/index.html) modules.
 #![deny(warnings)]
 #![deny(missing_debug_implementations, missing_copy_implementations)]
-#![recursion_limit = "128"]
+#![recursion_limit = "256"]
```
Comment (sgrif) [inline src/lib.rs]: Why was this necessary?

Comment (kivikakk) [reply src/lib.rs]: Without:

```
error: recursion limit reached while expanding the macro `table_body`
   --> src/schema.rs:223:1
    |
223 | / table! {
224 | |     use diesel_full_text_search::{TsVector as Tsvector};
225 | |     use diesel_ltree::Ltree;
226 | |     use diesel::sql_types::*;
...   |
322 | |     }
323 | | }
    | |_^
    |
    = help: consider adding a `#![recursion_limit="256"]` attribute to your crate
    = note: this error originates in a macro outside of the current crate

error: Could not compile `cargo-registry`.
```

File: src/category.rs (hunk 6 above)
Comment (kivikakk) [reply src/category.rs]: Done this (and fixed up the other tests too).

File: src/category.rs (hunk 5 above)
Comment (kivikakk) [reply src/category.rs]: :+1: Done.

**Comment by kivikakk** [comment]
```
error[E0063]: missing field `parent_categories` in initializer of `category::EncodableCategoryWithSubcategories`
   --> src/category.rs:526:19
    |
526 |         let cat = EncodableCategoryWithSubcategories {
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `parent_categories`
```

There is no line 526 of `category.rs`? The same line where it _actually_ exists does have a `parent_categories`? Travis what are you doing

**Comment by erewok** [comment]
nice work @kivikakk. I was actually wondering about this PR the other day. Thank you for continuing to push it forward.

File: src/models/category.rs [hunk 8]
```diff
@@ -93,43 +93,42 @@ impl Category {
...
         // the crates in all subcategories
-        select(sql::<categories::SqlType>(&format!(
-            "c.id, c.category, c.slug, c.description,
-                sum(c2.crates_cnt)::int as crates_cnt, c.created_at
-             FROM categories as c
-             INNER JOIN categories c2 ON split_part(c2.slug, '::', 1) = c.slug
-             WHERE split_part(c.slug, '::', 1) = c.slug
-             GROUP BY c.id
+        sql_query(format!(
+            "SELECT c.id, c.category, c.slug, c.description, \
```
Comment (sgrif) [inline src/models/category.rs]: How would you feel about moving this into a separate file and `include_str!`ing it. We can also change this to call `.bind` rather than using string interpolation (other than the order by)

Comment (kivikakk) [reply src/models/category.rs]: :+1:

**Comment by kivikakk** [comment]
Thank you for the ✅, sgrif!

Let me know if there's anything else to be done to get this merged, if we're still happy with it.

**Comment by carols10cents** [comment]
Thank you @erewok and @kivikakk for keeping this up to date, and sorry for the delay!

I tested this out and it looks like it's working great in the UI!

Buuuuut I noticed in the schema.rs that path was added as a column on the crates table, but if anything, it should be added to the categories table... but if I add it to the categories table by running `diesel print-schema --with-docs > src/schema.rs` and then manually adding `use diesel_ltree::Ltree;` within the `table!` call for categories (i.e. [this commit](https://github.com/rust-lang/crates.io/pull/1122/commits/fb335939b98f9e0646176bfe32b49b4553161bdd)) I then get errors, a bunch like this:

```
error[E0412]: cannot find type `Int4` in this scope
  --> src/schema.rs:83:15
   |
83 |         id -> Int4,
   |               ^^^^ did you mean `Into`?
```

So then I added `use diesel::sql_types::*;` [in this commit](https://github.com/rust-lang/crates.io/pull/1122/commits/60ef7ee2025471c16cdf0677d092c9323ff36f1c) to be more like the crates `table!`, which has custom types and works.

That gets rid of ~15 errors, but leaves me with 4 that look like this:

```
error[E0277]: the trait bound `(i32, std::string::String, std::string::String, std::string::String, i32, chrono::NaiveDateTime): diesel::Queryable<(diesel::sql_types::Integer, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Integer, diesel::sql_types::Timestamp, diesel_ltree::Ltree), _>` is not satisfied
  --> src/controllers/category.rs:42:10
   |
42 |         .first::<Category>(&*conn)?;
   |          ^^^^^ the trait `diesel::Queryable<(diesel::sql_types::Integer, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Integer, diesel::sql_types::Timestamp, diesel_ltree::Ltree), _>` is not implemented for `(i32, std::string::String, std::string::String, std::string::String, i32, chrono::NaiveDateTime)`
```

which, ok, the `Category` struct that's `Queryable` doesn't have the `path` column specified.

I tried adding `path` as type `LTree` in [this commit](https://github.com/rust-lang/crates.io/pull/1122/commits/be7fba190da706774e5640b19260a3435d3a952c), but then I get these errors:

```
error[E0277]: the trait bound `diesel_ltree::Ltree: diesel::Queryable<diesel_ltree::Ltree, diesel::pg::Pg>` is not satisfied
  --> src/controllers/category.rs:42:10
   |
42 |         .first::<Category>(&*conn)?;
   |          ^^^^^ the trait `diesel::Queryable<diesel_ltree::Ltree, diesel::pg::Pg>` is not implemented for `diesel_ltree::Ltree`
   |
```

The [tests in diesel_ltree use `Queryable` with the `path` column as type `String`](https://github.com/kivikakk/diesel_ltree/blob/master/src/tests.rs#L25), so I tried that too in [this commit](https://github.com/rust-lang/crates.io/pull/1122/commits/5986ee0c7d6f0953b1bf6bba18ce8c6d8e3fd63f), but then I get these errors:

```
error[E0277]: the trait bound `*const str: diesel::deserialize::FromSql<diesel_ltree::Ltree, diesel::pg::Pg>` is not satisfied
  --> src/controllers/category.rs:42:10
   |
42 |         .first::<Category>(&*conn)?;
   |          ^^^^^ the trait `diesel::deserialize::FromSql<diesel_ltree::Ltree, diesel::pg::Pg>` is not implemented for `*const str`
   |
   = help: the following implementations were found:
             <*const [u8] as diesel::deserialize::FromSql<diesel::sql_types::Binary, DB>>
             <*const str as diesel::deserialize::FromSql<diesel::sql_types::Text, DB>>
```

soooo @sgrif @kivikakk am I using diesel or diesel_ltree wrong? or is there something incompatible/missing here?

**Comment by carols10cents** [comment]
Ok, I think I've successfully figured this one out. It worked when the path column was accidentally on the crates table because the crates code was already explicitly selecting fewer than all columns; when the path column was added to the category table in the schema instead, the same don't-select-all-columns code needed to happen for category too.

@sgrif if you have a minute, could you review the last commit of mine and make sure i did this right?

**Comment by kivikakk** [comment]
It'd be nice if we could land this!

**Comment by carols10cents** [comment]
Yes, yes it would. YOLO

bors: r+

**Comment by carols10cents** [comment]
hMMMM why didn't that work....

bors: r+

**Comment by carols10cents** [comment]
there we go. good bot!

**Comment by carols10cents** [comment]
hmmmm looks like rustfmt had differences and clippy failed to install? i get different results when running rustfmt locally, fun. we'll see if this fixes it

**Comment by carols10cents** [comment]
ok let's try this one more time!

bors: r+

**Comment by carols10cents** [comment]
You're welcome! I'll try and get this deployed tomorrow :)

**Comment by erewok** [comment]
Wow, nice work all. I completely forgot about this PR (sorry about that) and found it just now in my github unread.
